BMEdit/Editor/Source/Widgets/SceneRenderWidget.cpp

```diff
@@ -993,218 +1016,302 @@ namespace widgets
-							gapi->glBlendFunc(GL_ONE, GL_ONE);
-							break;
-						case gamelib::mat::MATBlendMode::BM_ADD:
-							gapi->glBlendFunc(GL_ONE, GL_ONE);
-							break;
-						default:
-							// Do nothing
-							break;
-						}
-
-						// Enable or disable alpha testing
-						if (renderState.isAlphaTestEnabled()) {
-							gapi->glEnable(GL_ALPHA_TEST);
-						} else {
-							gapi->glDisable(GL_ALPHA_TEST);
+							assert(false && "Idk how to handle this");
+							continue;
 						}
 
-						// Enable or disable fog
-						if (renderState.isFogEnabled()) {
-							gapi->glEnable(GL_FOG);
-						} else {
-							gapi->glDisable(GL_FOG);
-						}
+						const auto& kind = texture.getName();
 
-						// Enable or disable depth offset (Z bias)
-						if (renderState.hasZBias()) {
-							gapi->glEnable(GL_POLYGON_OFFSET_FILL);
-							gapi->glPolygonOffset(1.0f, renderState.getZOffset());
-						} else {
-							gapi->glDisable(GL_POLYGON_OFFSET_FILL);
-						}
+						int textureSlotId = render::TextureSlotId::kMaxTextureSlot;
 
-						// Set cull mode based on your enum values
-						switch (renderState.getCullMode())
-						{
-                            case gamelib::mat::MATCullMode::CM_OneSided:
-                                gapi->glCullFace(GL_BACK);
-							break;
-                            case gamelib::mat::MATCullMode::CM_DontCare:
-						    case gamelib::mat::MATCullMode::CM_TwoSided:
-							    // please complete
-							    gapi->glDisable(GL_CULL_FACE);
-							    break;
-						}
-					};
+#define MATCH_TEXTURE_KIND(mode, modeName) if (kind == modeName) { textureSlotId = mode; }
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapDiffuse,           "mapDiffuse")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapSpecularMask,      "mapSpecularMask")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapEnvironment,       "mapEnvironment")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapReflectionMask,    "mapReflectionMask")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapReflectionFallOff, "mapReflectionFallOff")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapIllumination,      "mapIllumination")
+						MATCH_TEXTURE_KIND(render::TextureSlotId::kMapTranslucency,      "mapTranslucency")
+#undef MATCH_TEXTURE_KIND
 
-					// 0. Check that we've able to draw it
-					if (mesh.glTextureId == kInvalidResource)
-					{
-						// Draw "error" bounding box
-						// And continue
-						continue;
-					}
+						if (textureSlotId == render::TextureSlotId::kMaxTextureSlot)
+							continue;
 
-					// Filter mesh by 'variation id'
-					if (const auto& properties = entry.pGeom->getProperties(); properties.hasProperty("MeshVariantId"))
-					{
-						const auto requiredVariationId = properties.getObject<std::int32_t>("MeshVariantId", 0);
-						if (requiredVariationId != mesh.variationId)
+						// Now we need to find texture instance and associate it
+						for (const auto& g1tex : m_resources->m_textures)
```

Comment:
Need optimize this. Lookup by texture 'index' or 'name' on each 'build render list' query looks very hard.